src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseBuildConfigService.java

```diff
@@ -0,0 +1,161 @@
+     */
+    @Nullable
+    DockerRunConfig getDockerRunConfigFromParsedList(List<List<String>> list) {
+        final int keyIndex = 0;
+        final int valueIndex = 1;
+        try {
+            boolean networkDisabled = false;
+            List<String> env = null;
+            for (List<String> entry : list) {
+                if (entry.size() != 2 || StringUtils.isBlank(entry.get(valueIndex)) || StringUtils.isBlank(entry.get(keyIndex))
+                        || !DockerRunConfig.AllowedDockerFlags.isAllowed(entry.get(keyIndex))) {
+                    log.warn("Invalid Docker flag entry: {}. Skipping.", entry);
+                    continue;
+                }
+                switch (entry.get(keyIndex)) {
+                    case "network":
+                        networkDisabled = entry.get(valueIndex).equalsIgnoreCase("none");
+                        break;
+                    case "env":
+                        env = parseEnvVariableString(entry.get(valueIndex));
+                        break;
+                    default:
+                        log.error("Invalid Docker flag entry: {}. Skipping.", entry);
+                        break;
+                }
+
+            }
+            return new DockerRunConfig(networkDisabled, env);
+        }
+        catch (Exception e) {
+            log.error("Failed to parse DockerRunConfig from JSON string: {}. Using default settings.", list, e);
+        }
+
+        return null;
+    }
+
+    /**
+     * Parses the JSON string representing Docker flags into a list of key-value pairs.
+     *
+     * @return a list of key-value pairs, or {@code null} if an error occurs
+     */
+    @Nullable
+    List<List<String>> parseDockerFlags(ProgrammingExerciseBuildConfig buildConfig) {
+        if (StringUtils.isBlank(buildConfig.getDockerFlags())) {
+            return null;
+        }
+
+        try {
+            ObjectMapper objectMapper = new ObjectMapper();
+
+            return objectMapper.readValue(buildConfig.getDockerFlags(), new TypeReference<>() {
+            });
+        }
+        catch (Exception e) {
+            log.error("Failed to parse DockerRunConfig from JSON string: {}. Using default settings.", buildConfig.getDockerFlags(), e);
+        }
+
+        return null;
+    }
+
+    private List<String> parseEnvVariableString(String envVariableString) {
+        Pattern pattern = Pattern.compile(
+                // match key-value pairs, where the key can be a single word or a string in single or double quotes
+                // key-value pairs are separated by commas
+                "(?:'([^']+)'|\"([^\"]+)\"|(\\w+))=(?:'([^']*)'|\"([^\"]*)\"|([^,]+))");
+
+        if (envVariableString.length() > MAX_ENVIRONMENT_VARIABLES_DOCKER_FLAG_LENGTH) {
+            log.warn("The environment variables string is too long. It will be truncated to {} characters.", MAX_ENVIRONMENT_VARIABLES_DOCKER_FLAG_LENGTH);
```

Comment:
⚠️ This is potentially dangerous: What happens in case one of the last environment variables configures some security option that restricts what students are allowed to do?
By cutting it off here silently without information for the user, this option may not actually be configured for the build. The user would only (hopefully) notice this when editing the build script again at some later point.

Is there a reason to truncate it? The database should be able to store text of pretty much arbitrary length in the database.

---

Edit: The other class also checks the length and throws a user-visible warning. Question remains why this restriction is necessary.

Reply:
This was added as a workaround for a CodeQL vulnerability that didnt detect that the string is restricted to 1000 chars before